[4498] Add bulk getSandboxItems API calls

aachavarria opened the pull request:
https://github.com/craftercms/craftercms/issues/4498

aachavarria committed:
adding fetchItemsByPath service

rart commented on ui/app/src/services/content.ts:
```diff
@@ -807,23 +812,14 @@ export function getChildrenByPath(
-        site,
-        path === '/site/website' ? (options?.skipHomePathOverride ? path : '/site/website/index.xml') : path
-      )
+export function fetchItemsByPath(siteId: string, paths: string[], preferContent = true): Observable<DetailedItem[]> {
```
LLeave the third arg as an object of options. Default it to true. Create an interface just like with get children.

rart commented on ui/app/src/services/content.ts:
```diff
@@ -807,23 +812,14 @@ export function getChildrenByPath(
+  return get(`/studio/api/2/content/sandbox_items_by_path${qs}`).pipe(
+    pluck('response', 'items'),
+    map((items) =>
+      items.map((item) => ({ ...parseSandBoxItemToDetailedItem(item), stateMap: createItemStateMap(item.state) }))
```
Add options.castAsDetailedItem and only apply this if such option is true. Try to create the function overloads to specify when it returns Detailed or Sandbox

aachavarria committed:
PR Feedback;

aachavarria committed:
Overload

rart commented on ui/app/src/services/content.ts:
```diff
@@ -807,23 +813,29 @@ export function getChildrenByPath(
+  siteId: string,
+  paths: string[],
+  options?: FetchItemsByPathOptions
+): Observable<SandboxItem[] | DetailedItem[]> {
```
Typings not sufficient for correct type deduction.

rart committed:
Improve type inflection cues

rart committed:
Fix order of overloads

Curiously, the order of overloads affects the type deduction. Fixed order to produce correct deductions.

rart commented on ui/app/src/services/content.ts:
```diff
@@ -807,23 +813,39 @@ export function getChildrenByPath(
+  paths: string[],
+  options?: FetchItemsByPathOptions
+): Observable<SandboxItem[] | DetailedItem[]> {
+  const { castAsDetailedItem = false, preferContent = true } = options;
```
Need to protect against options being undefined (i.e. can't restructure properties from undefined)
```suggestion
  const { castAsDetailedItem = false, preferContent = true } = options ?? {};
```

rart committed:
Protect against options being undefined
